### PR TITLE
IPv4 addresses, convert config to openstruct

### PR DESCRIPTION
**Pull request by fishnix**

### CODE REVIEW

**Comment by tenyo** [comment]
Should update the URL in the README:
```
gem 'proteus', :git => 'git://git.yale.edu/inf-sa/proteus_client.git'
```

File: exe/proteus [hunk 1]
```diff
@@ -74,79 +81,252 @@ command :info do |c|
...
+          # TODO: paging is required here -- probably need a helper
+          @proteus.get_zones_by_hint(config_id, 0, 10, "hint=#{hint}").each do |z|
+            begin
+              puts @proteus.get_entity_by_name(z.id, name, Proteus::Types::HOSTRECORD)
+            rescue Proteus::ApiEntityError::EntityNotFound
+              puts "#{name} not found in zone #{hint} with id #{z.id}"
+            end
+          end
+        end
+      end
```
Comment (tenyo) [inline exe/proteus:123]: image

Comment (fishnix) [reply exe/proteus:123]: lol yeah i had the same reaction, its mostly the GLI DSL though that's causing it....

**Comment by tenyo** [comment]
There seems quite a bit of logic in the CLI - I wonder if it can be moved into some complex actions methods, so it can be used for other things (API).

File: exe/proteus [hunk 2]
```diff
@@ -74,79 +81,252 @@ command :info do |c|
...
+  h.arg_name 'fqdn'
+  h.command :show do |s|
+    s.action do |_global_options, _options, args|
       validate_argnum(1, args.length)
-      ap @proteus.search_by_object_types(args[0], Proteus::Types::EXTERNALHOST)
+      config_id = @proteus.get_entities.first.id
+      collect_pairs(args[0].split '.').each do |name, hint|
+        unless name == '' || hint == ''
+          # TODO: paging is required here -- probably need a helper
+          @proteus.get_zones_by_hint(config_id, 0, 10, "hint=#{hint}").each do |z|
```
Comment (tenyo) [inline exe/proteus:115]: Do we wanna stop searching as soon as we find an entity match, or go through all the zones regardless? Probably the first as the host record should be unique.

File: lib/proteus/actions/entity.rb [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+      # deletes an object by id
+      # <message name="ProteusAPI_delete">
+      #   <part name="objectId" type="xsd:long"/>
+      # </message>
+      def delete(id)
+        entity = get_entity_by_id(id)
+        if ALLOWDELETE.include?(entity.type)
+          call(:delete, object_id: id)
+        else
+          raise 'Not allowed to delete entity with type: ' + entity.type
```
Comment (tenyo) [inline lib/proteus/actions/entity.rb]: Should this raise a custom error class?

File: exe/proteus [hunk 4]
```diff
@@ -74,79 +81,252 @@ command :info do |c|
...
+    a.action do |_global_options, options, args|
       validate_argnum(2, args.length)
-      ap @proteus.add_host_record(args[0], args[1])
+      config_id = @proteus.get_entities.first.id
+
+      # If we are passed an IP, use it.  Otherwise generate a canary.
+      # Note: it's _possible_ the first IP in a net block is unallocated, so
+      # there could be an edge case where this fails.
+      if options[:useip] || args[0].end_with?('/32')
+        canary_ip = NetAddr::CIDR.create(args[0]).first
```
Comment (tenyo) [inline exe/proteus:223]: Looks like `first` is an alias for `network`, which may be less confusing here.

**Comment by tenyo** [comment]
This is awesome! Just the questions above. 👍

File: exe/proteus (hunk 4 above)
Comment (fishnix) [reply exe/proteus:223]: In this case I don't think so -- 'network' implies `x.x.x.0` for example, but this is a `/32`

File: exe/proteus (hunk 2 above)
Comment (fishnix) [reply exe/proteus:115]: Yeah, probably true.  Let's do that when we do paging.

**Comment by fishnix** [comment]
I think we should add some helpers to the gem, but writing the internal API might help inform that.  thanks for the review!

**Pull request merged**